econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -0,0 +1,1071 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
```

Comment:
The following raises a bug:
```python
from econml.solutions.causal_analysis import CausalAnalysis
import numpy as np
from itertools import product
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline

# Treatment effect function
def exp_te(X):
    return np.exp(2*X[:, 0])
# DGP constants
np.random.seed(123)
n = 5000
n_w = 30
support_size = 5
n_x = 2
# Outcome support
support_Y = np.random.choice(range(n_w), size=support_size, replace=False)
coefs_Y = np.random.uniform(0, 1, size=support_size)
epsilon_sample = lambda n: np.random.uniform(-1, 1, size=n)
# Treatment support 
support_T = support_Y
coefs_T = np.random.uniform(0, 1, size=support_size)
eta_sample = lambda n: np.random.uniform(-1, 1, size=n) 

# Generate controls, covariates, treatments and outcomes
W = np.random.normal(0, 1, size=(n, n_w))
X = np.random.uniform(0, 1, size=(n, n_x))
# Heterogeneous treatment effects
TE = exp_te(X)
Tind = np.dot(W[:, support_T], coefs_T) + eta_sample(n)
T = (Tind>0)
Y = TE * T + np.dot(W[:, support_Y], coefs_Y) + epsilon_sample(n)
T = np.empty(Tind.shape, dtype='str')
T[Tind>0] = 'a'
T[Tind<=0] = 'b'
# Y = (Y > 0)
X_test = np.array(list(product(np.arange(0, 1, 0.01), repeat=n_x)))
A = np.hstack([T.reshape(-1, 1), X, W])
Xcols = [f'X{i}' for i in range(n_x)] 
Wcols = [f'W{i}' for i in range(n_w)]
Adf = pd.DataFrame(A, columns=['T'] + Xcols + Wcols)
ca = CausalAnalysis(feature_inds=['T'],
                    heterogeneity_inds=[Xcols],
                    categorical=['T'],
                    classification=False)
ca.fit(Adf, Y)
ca.feature_inds = ['T', 'X0', f'W{support_Y[0]}']
ca.heterogeneity_inds = [Xcols, Xcols, Xcols]
ca.fit(Adf, Y, warm_start=True)
```

```
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\joblib\externals\loky\process_executor.py", line 431, in _process_worker
    r = call_item()
  File "C:\ProgramData\Anaconda3\lib\site-packages\joblib\externals\loky\process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\joblib\_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\joblib\parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\joblib\parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\solutions\causal_analysis\_causal_analysis.py", line 438, in process_feature
    est.fit(y, T, X=X_xf, W=W, cache_values=True)
  File "c:\users\vasy\documents\econml\econml\utilities.py", line 1262, in m
    return to_wrap(*args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\dml\dml.py", line 659, in fit
    return super().fit(Y, T, X=X, W=W,
  File "c:\users\vasy\documents\econml\econml\utilities.py", line 1262, in m
    return to_wrap(*args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\dml\dml.py", line 505, in fit
    return super().fit(Y, T, X=X, W=W, sample_weight=sample_weight, freq_weight=freq_weight,
  File "c:\users\vasy\documents\econml\econml\utilities.py", line 1262, in m
    return to_wrap(*args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\dml\_rlearner.py", line 369, in fit
    return super().fit(Y, T, X=X, W=W,
  File "c:\users\vasy\documents\econml\econml\utilities.py", line 1262, in m
    return to_wrap(*args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\_cate_estimator.py", line 130, in call
    m(self, Y, T, *args, **kwargs)
  File "c:\users\vasy\documents\econml\econml\_ortho_learner.py", line 638, in fit
    nuisances, fitted_models, new_inds, scores = self._fit_nuisances(
  File "c:\users\vasy\documents\econml\econml\_ortho_learner.py", line 767, in _fit_nuisances
    nuisances, fitted_models, fitted_inds, scores = _crossfit(self._ortho_learner_model_nuisance, folds,
  File "c:\users\vasy\documents\econml\econml\_ortho_learner.py", line 170, in _crossfit
    nuisance_temp = model_list[idx].predict(*args_test, **kwargs_test)
  File "c:\users\vasy\documents\econml\econml\dml\_rlearner.py", line 75, in predict
    T_res = T - T_pred.reshape(T.shape)
TypeError: unsupported operand type(s) for -: 'str' and 'float'
"""

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
<ipython-input-89-f52cbe5cafa7> in <module>
      2 ca.heterogeneity_inds = [Xcols, Xcols, Xcols]
      3 # ca.nuisance_models = 'automl'
----> 4 ca.fit(Adf, Y, warm_start=True)

c:\users\vasy\documents\econml\econml\solutions\causal_analysis\_causal_analysis.py in fit(self, X, y, warm_start)
    500 
    501         cache_updates = dict(zip(new_inds,
--> 502                                  joblib.Parallel(n_jobs=self.n_jobs,
    503                                                  verbose=1)(joblib.delayed(process_feature)(feat_name, feat_ind)
    504                                                             for feat_name, feat_ind in zip(new_feat_names, new_inds))))

C:\ProgramData\Anaconda3\lib\site-packages\joblib\parallel.py in __call__(self, iterable)
   1059 
   1060             with self._backend.retrieval_context():
-> 1061                 self.retrieve()
   1062             # Make sure that we get a last message telling us we are done
   1063             elapsed_time = time.time() - self._start_time

C:\ProgramData\Anaconda3\lib\site-packages\joblib\parallel.py in retrieve(self)
    938             try:
    939                 if getattr(self._backend, 'supports_timeout', False):
--> 940                     self._output.extend(job.get(timeout=self.timeout))
    941                 else:
    942                     self._output.extend(job.get())

C:\ProgramData\Anaconda3\lib\site-packages\joblib\_parallel_backends.py in wrap_future_result(future, timeout)
    540         AsyncResults.get from multiprocessing."""
    541         try:
--> 542             return future.result(timeout=timeout)
    543         except CfTimeoutError as e:
    544             raise TimeoutError from e

C:\ProgramData\Anaconda3\lib\concurrent\futures\_base.py in result(self, timeout)
    437                 raise CancelledError()
    438             elif self._state == FINISHED:
--> 439                 return self.__get_result()
    440             else:
    441                 raise TimeoutError()

C:\ProgramData\Anaconda3\lib\concurrent\futures\_base.py in __get_result(self)
    386     def __get_result(self):
    387         if self._exception:
--> 388             raise self._exception
    389         else:
    390             return self._result

TypeError: unsupported operand type(s) for -: 'str' and 'float'
```

Reply:
I think when you hstack the columns, the resulting array has a weird type because the first column has strings in it, so things get messed up from there.